app/models/assignment.rb

```diff
@@ -850,7 +853,67 @@ def get_num_valid
              .count(&:is_valid?)
   end
 
-  def get_num_marked(ta_id = nil)
+  def marked_result_ids_for(ta_id)
+    cache_ta_results
+    @cache_ta_results.dig(ta_id, :marked_result_ids) || []
+  end
+
+  def cache_ta_results
+    return @cache_ta_results if defined? @cache_ta_results
+
+    data = current_results.joins(grouping: :tas).pluck('tas.id', 'results.id', 'results.marking_state')
+    # Group results by TA ID
+    grouped_data = data.group_by { |ta_id, _result_id, _marking_state| ta_id }
+
+    results_by_id = current_results.includes(:marks).index_by(&:id)
+    @cache_ta_results = grouped_data.each_with_object({}) do |(ta_id, results), hash|
```

Comment:
You don't need to use `each_with_object` here. Instead, you can use `Hash.[]` and inside pass a list of lists, where each inner list is of the form `[key, value]`. There are a few other places in the MarkUs codebase that do this, but as a hint the overall structure is something like

```ruby
@cache_ta_results = Hash[grouped_data.map |ta_id, results| do
  ...
  end
]
```